Update TinkerPop to 3.4.5

TinkerPop 3.4.5 is released.

Just bumping version from `3.4.4` to `3.4.5` doesn't work for us because TinkerPop tests fail for JanusGraph when changing version from `3.4.4` to `3.4.5`.

Tests are executed with: `mvn clean install -Dtest.skip.tp=false -DskipTests=true`

```
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.janusgraph.blueprints.inmemory.process.InMemoryJanusGraphComputerTest
[ERROR] Tests run: 741, Failures: 2, Errors: 0, Skipped: 141, Time elapsed: 820.182 s <<< FAILURE! - in org.janusgraph.blueprints.inmemory.process.InMemoryJanusGraphComputerTest
[ERROR] g_V_bothE_properties_dedup_hasKeyXweightX_hasValueXltX0d3XX_value  Time elapsed: 0.513 s  <<< FAILURE!
java.lang.AssertionError: Checking result size expected:<1> but was:<0>

[ERROR] g_V_bothE_properties_dedup_hasKeyXweightX_value  Time elapsed: 0.504 s  <<< FAILURE!
java.lang.AssertionError: Checking result size expected:<4> but was:<0>

[INFO] Running org.janusgraph.blueprints.inmemory.process.InMemoryJanusGraphProcessTest
[ERROR] Tests run: 782, Failures: 2, Errors: 0, Skipped: 19, Time elapsed: 179.068 s <<< FAILURE! - in org.janusgraph.blueprints.inmemory.process.InMemoryJanusGraphProcessTest
[ERROR] g_V_bothE_properties_dedup_hasKeyXweightX_hasValueXltX0d3XX_value  Time elapsed: 0.219 s  <<< FAILURE!
java.lang.AssertionError: Checking result size expected:<1> but was:<0>

[ERROR] g_V_bothE_properties_dedup_hasKeyXweightX_value  Time elapsed: 0.223 s  <<< FAILURE!
java.lang.AssertionError: Checking result size expected:<4> but was:<0>

[INFO] Running org.janusgraph.blueprints.inmemory.process.InMemoryMultiQueryJanusGraphProcessTest
[ERROR] Tests run: 782, Failures: 2, Errors: 0, Skipped: 19, Time elapsed: 180.396 s <<< FAILURE! - in org.janusgraph.blueprints.inmemory.process.InMemoryMultiQueryJanusGraphProcessTest
[ERROR] g_V_bothE_properties_dedup_hasKeyXweightX_hasValueXltX0d3XX_value  Time elapsed: 0.218 s  <<< FAILURE!
java.lang.AssertionError: Checking result size expected:<1> but was:<0>

[ERROR] g_V_bothE_properties_dedup_hasKeyXweightX_value  Time elapsed: 0.216 s  <<< FAILURE!
java.lang.AssertionError: Checking result size expected:<4> but was:<0>

[INFO] Running org.janusgraph.blueprints.inmemory.structure.InMemoryJanusGraphStructureTest
[WARNING] Tests run: 934, Failures: 0, Errors: 0, Skipped: 266, Time elapsed: 135.152 s - in org.janusgraph.blueprints.inmemory.structure.InMemoryJanusGraphStructureTest
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR] org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest$Traversals.g_V_bothE_properties_dedup_hasKeyXweightX_hasValueXltX0d3XX_value
[ERROR]   Run 1: HasTest$Traversals>HasTest.g_V_bothE_properties_dedup_hasKeyXweightX_hasValueXltX0d3XX_value:497->AbstractGremlinProcessTest.checkResults:134 Checking result size expected:<1> but was:<0>
[ERROR]   Run 2: HasTest$Traversals>HasTest.g_V_bothE_properties_dedup_hasKeyXweightX_hasValueXltX0d3XX_value:497->AbstractGremlinProcessTest.checkResults:134 Checking result size expected:<1> but was:<0>
[ERROR]   Run 3: HasTest$Traversals>HasTest.g_V_bothE_properties_dedup_hasKeyXweightX_hasValueXltX0d3XX_value:497->AbstractGremlinProcessTest.checkResults:134 Checking result size expected:<1> but was:<0>
[INFO] 
[ERROR] org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest$Traversals.g_V_bothE_properties_dedup_hasKeyXweightX_value
[ERROR]   Run 1: HasTest$Traversals>HasTest.g_V_bothE_properties_dedup_hasKeyXweightX_value:489->AbstractGremlinProcessTest.checkResults:134 Checking result size expected:<4> but was:<0>
[ERROR]   Run 2: HasTest$Traversals>HasTest.g_V_bothE_properties_dedup_hasKeyXweightX_value:489->AbstractGremlinProcessTest.checkResults:134 Checking result size expected:<4> but was:<0>
[ERROR]   Run 3: HasTest$Traversals>HasTest.g_V_bothE_properties_dedup_hasKeyXweightX_value:489->AbstractGremlinProcessTest.checkResults:134 Checking result size expected:<4> but was:<0>
[INFO] 
[INFO] 
[ERROR] Tests run: 3070, Failures: 2, Errors: 0, Skipped: 280
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] JanusGraph: Distributed Graph Database 0.5.0-SNAPSHOT SUCCESS [  0.798 s]
[INFO] JanusGraph-Driver: Gremlin-Driver Library for JanusGraph SUCCESS [  1.596 s]
[INFO] JanusGraph-Core: Core Library for JanusGraph ....... SUCCESS [ 11.327 s]
[INFO] JanusGraph-Backend-testutils: Collection of Backend tests SUCCESS [  3.232 s]
[INFO] JanusGraph CodePipelines CI: Distributed release testing. SUCCESS [  2.884 s]
[INFO] janusgraph-server .................................. SUCCESS [  0.509 s]
[INFO] JanusGraph-InMemory: InMemory Backend for JanusGraph FAILURE [22:02 min]
[INFO] JanusGraph-Test: Test Suite for JanusGraph ......... SKIPPED
[INFO] JanusGraph-BerkeleyJE: Distributed Graph Database .. SKIPPED
[INFO] JanusGraph-Hadoop: Module .......................... SKIPPED
[INFO] JanusGraph-CQL: Distributed Graph Database ......... SKIPPED
[INFO] JanusGraph-Cassandra: Distributed Graph Database ... SKIPPED
[INFO] JanusGraph-HBase: Parent Module .................... SKIPPED
[INFO] JanusGraph-HBase: Version-independent Core ......... SKIPPED
[INFO] JanusGraph-HBase: HBase Server Used by Maven Testing SKIPPED
[INFO] JanusGraph-HBase: 1.x and 2.x Compatibility Shim ... SKIPPED
[INFO] JanusGraph-HBase: Universal binary ................. SKIPPED
[INFO] JanusGraph-Bigtable: Google Cloud Bigtable driver .. SKIPPED
[INFO] JanusGraph-ElasticSearch: Distributed Indexing Support SKIPPED
[INFO] JanusGraph-Lucene: Indexing Support ................ SKIPPED
[INFO] JanusGraph-Solr: Distributed Indexing Support ...... SKIPPED
[INFO] JanusGraph-All: Complete JanusGraph Distribution ... SKIPPED
[INFO] JanusGraph-Dist: Tar and Zip Archives .............. SKIPPED
[INFO] JanusGraph-Doc: Manual for JanusGraph .............. SKIPPED
[INFO] JanusGraph-Examples: Examples for JanusGraph ....... SKIPPED
[INFO] Example-Common: Common Graph Code for Examples ..... SKIPPED
[INFO] Example-BerkeleyJE: BerkeleyJE Storage, Lucene Index SKIPPED
[INFO] Example-Cassandra: Cassandra Thrift Storage, Elasticsearch Index SKIPPED
[INFO] Example-Cql: Cassandra CQL Storage, Elasticsearch Index SKIPPED
[INFO] Example-HBase: HBase Storage, Solr Index ........... SKIPPED
[INFO] Example-RemoteGraph: Example with RemoteGraph ...... SKIPPED
[INFO] Example-TinkerGraph: Example with TinkerGraph 0.5.0-SNAPSHOT SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 22:23 min
[INFO] Finished at: 2020-02-07T22:44:53-08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.1:test (tinkerpop-test) on project janusgraph-inmemory: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/ws/janusgraph/janusgraph-inmemory/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :janusgraph-inmemory
```

--- a/docs/changelog.md
+++ b/docs/changelog.md
@@ -50,7 +50,7 @@ compile "org.janusgraph:janusgraph-core:0.5.0"
 * Elasticsearch 6.0.1, 6.6.0, 7.3.1
 * Apache Lucene 7.0.0
 * Apache Solr 7.0.0
-* Apache TinkerPop 3.4.4
+* Apache TinkerPop 3.4.5
 * Java 1.8
 
 For more information on features and bug fixes in 0.5.0, see the GitHub milestone:

--- a/janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
+++ b/janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
@@ -4171,7 +4171,6 @@ public abstract class JanusGraphTest extends JanusGraphBaseTest {
 
     }
 
-
     @Test
     public void testSimpleTinkerPopTraversal() {
         Vertex v1 = graph.addVertex("name", "josh");
@@ -4183,6 +4182,47 @@ public abstract class JanusGraphTest extends JanusGraphBaseTest {
         assertNotNull(id);
     }
 
+    @Test
+    public void testHasKeyOnEdgePropertyTraversal() {
+
+        Vertex vertex = prepareDataForEdgePropertyFilterTest();
+
+        List<Object> result = graph.traversal().V(vertex).bothE().properties().hasKey("name").value().toList();
+        assertEquals(1, result.size());
+        assertEquals(result.get(0), "testValue");
+    }
+
+    @Test
+    public void testHasValueOnEdgePropertyTraversal() {
+
+        Vertex vertex = prepareDataForEdgePropertyFilterTest();
+
+        List<Object> result = graph.traversal().V(vertex).bothE().properties().hasValue("testValue").value().toList();
+        assertEquals(1, result.size());
+        assertEquals(result.get(0), "testValue");
+    }
+
+    @Test
+    public void testHasKeyAndHasValueOnEdgePropertyTraversal() {
+
+        Vertex vertex = prepareDataForEdgePropertyFilterTest();
+
+        List<Object> result = graph.traversal().V(vertex).bothE().properties().hasKey("weight").hasValue(P.lt(3)).value().toList();
+        assertEquals(1, result.size());
+        assertEquals(result.get(0), 2);
+    }
+
+    private Vertex prepareDataForEdgePropertyFilterTest(){
+
+        Vertex v1 = graph.addVertex("name", "josh");
+        Vertex v2 = graph.addVertex("name", "lop");
+        Vertex v3 = graph.addVertex("name", "lop2");
+
+        graph.traversal().V(v1).addE("created").to(v2).property("name", "testValue").next();
+        graph.traversal().V(v1).addE("created").to(v3).property("weight", 2).next();
+
+        return v1;
+    }
 
    /* ==================================================================================
                             LOGGING

--- a/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphPropertiesStep.java
+++ b/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphPropertiesStep.java
@@ -17,7 +17,6 @@ package org.janusgraph.graphdb.tinkerpop.optimize;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Iterators;
-import org.apache.tinkerpop.gremlin.structure.Property;
 import org.janusgraph.core.*;
 import org.janusgraph.graphdb.query.BaseQuery;
 import org.janusgraph.graphdb.query.JanusGraphPredicateUtils;
@@ -33,6 +32,7 @@ import org.apache.tinkerpop.gremlin.process.traversal.step.util.HasContainer;
 import org.apache.tinkerpop.gremlin.process.traversal.util.FastNoSuchElementException;
 import org.apache.tinkerpop.gremlin.process.traversal.util.MutableMetrics;
 import org.apache.tinkerpop.gremlin.structure.Element;
+import org.apache.tinkerpop.gremlin.structure.Property;
 import org.apache.tinkerpop.gremlin.structure.Vertex;
 import org.apache.tinkerpop.gremlin.structure.util.StringFactory;
 import org.apache.tinkerpop.gremlin.structure.util.wrapped.WrappedVertex;
@@ -104,10 +104,10 @@ public class JanusGraphPropertiesStep<E> extends PropertiesStep<E> implements Ha
      * an attempt to retrieve a vertex from the cached multiQuery results doesn't find an entry.
      * @param vertices A list of vertices with which to initialise the multiQuery
      */
-    private void initializeMultiQuery(final List<Traverser.Admin<Element>> list) {
-        assert list.size() > 0;
+    private void initializeMultiQuery(final List<Traverser.Admin<Element>> vertices) {
+        assert vertices.size() > 0;
         final JanusGraphMultiVertexQuery multiQuery = JanusGraphTraversalUtil.getTx(traversal).multiQuery();
-        list.forEach(v -> multiQuery.addVertex((Vertex)v.get()));
+        vertices.forEach(v -> multiQuery.addVertex((Vertex)v.get()));
         makeQuery(multiQuery);
 
         Map<JanusGraphVertex, Iterable<? extends JanusGraphProperty>> results = multiQuery.properties();
@@ -144,9 +144,16 @@ public class JanusGraphPropertiesStep<E> extends PropertiesStep<E> implements Ha
             } else {
                 //this asks for properties
                 assert orders.isEmpty();
-                //HasContainers don't apply => empty result set
-                if (!hasContainers.isEmpty()) return Collections.emptyIterator();
                 iterator = (Iterator<E>) traverser.get().properties(getPropertyKeys());
+                if (!hasContainers.isEmpty()) {
+                    List<E> properties = new LinkedList<>();
+                    iterator.forEachRemaining(e -> {
+                        if(HasContainer.testAll(e, hasContainers)){
+                            properties.add(e);
+                        }
+                    });
+                    iterator = properties.iterator();
+                }
             }
             if (limit!=Query.NO_LIMIT) iterator = Iterators.limit(iterator,limit);
             return iterator;

--- a/mkdocs.yml
+++ b/mkdocs.yml
@@ -127,6 +127,6 @@ nav:
 extra:
     latest_version: 0.5.0
     snapshot_version: 0.5.0-SNAPSHOT
-    tinkerpop_version: 3.4.4
+    tinkerpop_version: 3.4.5
     hadoop2_version: 2.7.7
     jamm_version: 0.3.0

--- a/pom.xml
+++ b/pom.xml
@@ -69,7 +69,7 @@
     </distributionManagement>
     <properties>
         <titan.compatible-versions>1.0.0,1.1.0-SNAPSHOT</titan.compatible-versions>
-        <tinkerpop.version>3.4.4</tinkerpop.version>
+        <tinkerpop.version>3.4.5</tinkerpop.version>
         <junit-platform.version>1.5.1</junit-platform.version>
         <junit.version>5.5.1</junit.version>
         <mrunit.version>1.1.0</mrunit.version>

